frontend/src/api/opsAPI.js

```diff
@@ -24,6 +24,76 @@ export const opsApi = createApi({
         }),
         getAgreementById: builder.query({
             query: (id) => `/agreements/${id}`,
+            providesTags: ["Agreements"],
+        }),
+        addAgreement: builder.mutation({
+            query: (data) => {
+                // remove fields that are not allowed
+                // eslint-disable-next-line no-unused-vars
+                const { id, budget_line_items, created_by, created_on, updated_on, ...postData } = data;
+                return {
+                    url: `/agreements/`,
+                    method: "POST",
+                    headers: { "Content-Type": "application/json" },
+                    body: { ...postData, number: "" },
+                };
+            },
+            invalidatesTags: ["Agreements", "BudgetLineItems"],
+        }),
+        updateAgreement: builder.mutation({
+            query: ({ id, data }) => {
+                // remove fields that are not allowed
+                // eslint-disable-next-line no-unused-vars
+                const { id: _id, budget_line_items, created_by, created_on, updated_on, ...patchData } = data;
+                return {
+                    url: `/agreements/${id}`,
+                    method: "PATCH",
+                    headers: { "Content-Type": "application/json" },
+                    body: patchData,
+                };
+            },
+            invalidatesTags: ["Agreements", "BudgetLineItems"],
+        }),
+        addBudgetLineItem: builder.mutation({
+            query: ({ data }) => {
+                const cleanData = { ...data }; // make a copy
+                if (cleanData.date_needed === "--") {
+                    cleanData.date_needed = null;
+                }
+                delete cleanData.created_by;
+                delete cleanData.can;
+                delete cleanData.id;
+
+                return {
+                    url: `/budget-line-items/`,
+                    method: "POST",
+                    headers: { "Content-Type": "application/json" },
+                    body: cleanData,
+                };
+            },
+            invalidatesTags: ["Agreements", "BudgetLineItems"],
+        }),
+        updateBudgetLineItem: builder.mutation({
+            query: ({ data }) => {
+                const cleanData = { ...data }; // make a copy
```

Comment:
The problem with this approach is that it makes assumptions about `data`, i.e. if another part of the app wanted to call `updateBudgetLineItem` it would be forced to pass in `data` in this particular format (which is awkward).  It would be better to normalize `data` in the callers at not here IMHO.  We can talk more in office hours on Monday.

Reply:
Not blocking the PR - this came up earlier too when Frank mentioned in Slack about discussing `opsAPI`

Reply:
Ok, we can merge this and discuss on Monday and do a small follow up PR.
The date_needed === "--" -> null seems like it doesn't belong in the API, but some of the cleaning is actually  removing fields that were in a GET that aren't allow in a PUT/POST, so that might be different.